src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -1606,5 +1606,111 @@ public static bool Overlaps<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> other,
                 value, comparer);
             return BinarySearch(span, comparable);
         }
+
+        /// <summary>
+        /// Sorts the elements in the entire <see cref="Span{T}" /> 
+        /// using the <see cref="IComparable" /> implementation of each 
+        /// element of the <see cref= "Span{T}" />
+        /// </summary>
+        /// <param name="span">The <see cref="Span{T}"/> to sort.</param>
+        /// <exception cref = "InvalidOperationException"> 
+        /// One or more elements do not implement the <see cref="IComparable" /> interface.
+        /// </exception>
+        public static void Sort<T>(this Span<T> span) =>
+            Sort(span, (IComparer<T>)null!);
+
+        /// <summary>
+        /// Sorts the elements in the entire <see cref="Span{T}" /> 
+        /// using the <typeparamref name="TComparer" />.
+        /// </summary>
+        public static void Sort<T, TComparer>(this Span<T> span, TComparer comparer)
```

Comment:
This should be `TComparer? comparer` instead of `TComparer`.

Reply:
Is this actually the desired / approved shape, having Sort be generic on both the T and the comparer type?

Reply:
> approved shape, having Sort be generic on both the T and the comparer type?

@stephentoub yes, this is the approved API. It was specifically designed to allow for value type comparers and hence specific code paths for that. Yes that means "code bloat" that is the purpose of it. :) In fact, it was a specific design goal of mine for the API after struggling for years with the restrictive and somewhat inconsistent `Array.Sort` API, where we have often had to simply  resort to put `stl::sort` in a native dll and interop to it, given we had native memory and a memory copy would be too expensive, and had specific sorting needs.

There already is specific code paths and "code bloat" for value types in current sorting code for `IComparable<T>` where `T : struct`, nothing new here, this simply adds orthogonality to it.

Reply:
> It was specifically designed to allow for value type comparers and hence specific code paths for that

Is that really a common case?  This implementation also doesn't appear to be benefiting from that at all, and worse, hides the fact that it's allocating.

Reply:
> Is that really a common case? 

@stephentoub in my view yes. I have seen enough `Array.Reverse` too. In my view there is no downside to this, most users will simply use `Sort()` or `Sort((x, y) => x.CompareTo(y))`, advanced users who need this can use the value type comparer.

> doesn't appear to be benefiting from that at all, and worse, hides the fact that it's allocating.

@stephentoub we agreed to do this step wise. One step at a time. I intend to fix this in later PRs, this will require code duplication etc.

Reply:
@stephentoub note that `Array.Sort` already does allocs for a custom comparer, converting it to `Comparison<T>`. I have noted this before and shown this in benchmarks. Can't run the latest ones now since I can't build coreclr anymore.

Reply:
I think it would be best to omit this shape of the API from this step.

The value prop of this special API is better performance. If the implementation does not actually provide better performance, it is just going to cause confusion.